### DISCUSSION TITLE
Question on PINO Implementation for BIF-to-Temperature Operator

### DISCUSSION THREAD

**Discussion by JarifulHassan** [lululxvi/deepxde#2054] [Q&A]
Dear DeepXDE Team,

I am using DeepXDE to learn an **operator mapping from a BIF (Bi_u : Biot number) to the transient temperature field** in a 2D heat conduction problem.

The setup is as follows:

* **Domain:** 2D rectangle ([0,L_x] × [0,L_y]) with time t ∈ [0,t_f]
* **PDE:** du/dt = d²u/dx² + d²u/dy²
* **Boundary conditions:**

  * Left and bottom: Neumann (∂u/∂n = 0)
  * Right: Robin (∂u/∂x + Bi_r * u = 0)
  * Top: Robin (∂u/∂y + Bi_u * u = 0)
* **Initial condition:** u(x,y,0) = 1

I am implementing a **physics-informed DeepONet (PINO)** using DeepXDE:

* I use **`TimePDE`** to generate the **training data** domain points, boundary points, and initial points and compute the PDE and BC residuals.
* `PDEOperator` with `PowerSeries` function space is used to represent the **BIF** top one , which is the **input function to the operator**.
* `DeepONet` branch-trunk network maps the BIF to the solution field u(x,y,t).
* Output transform enforces the initial condition.
* The **loss is fully physics-informed**, including PDE + boundary residuals.
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Minimal reproducible PINO example for 2D heat conduction with DeepXDE
Author: Jariful Hassan
"""

import tensorflow.compat.v1 as tf
tf.disable_v2_behavior()
import deepxde as dde
import numpy as np

### -------------------------------
Lx, Ly = 1.0, 1.0
t_f = 1.0

geom = dde.geometry.Rectangle([0, 0], [Lx, Ly])
timedomain = dde.geometry.TimeDomain(0, t_f)
geomtime = dde.geometry.GeometryXTime(geom, timedomain)

### -------------------------------
def pde(x, u, Bi_f):
    u_t = dde.grad.jacobian(u, x, i=0, j=2)
    u_xx = dde.grad.hessian(u, x, i=0, j=0)
    u_yy = dde.grad.hessian(u, x, i=1, j=1)
    return u_t - u_xx - u_yy

### -------------------------------
def boundary_l(x, on_boundary):
    return on_boundary and np.isclose(x[0], 0.0)

def boundary_r(x, on_boundary):
    return on_boundary and np.isclose(x[0], Lx)

def boundary_d(x, on_boundary):
    return on_boundary and np.isclose(x[1], 0.0)

def boundary_u(x, on_boundary):
    return on_boundary and np.isclose(x[1], Ly)

### Neumann BC left & bottom
bc_l = dde.icbc.NeumannBC(geomtime, lambda x: 0.0, boundary_l)
bc_d = dde.icbc.NeumannBC(geomtime, lambda x: 0.0, boundary_d)

### Robin BC right & top
bc_r = dde.icbc.OperatorBC(geomtime, lambda x, u, _: dde.grad.jacobian(u, x, i=0, j=0) + Bi_r * u, boundary_r)
bc_u = dde.icbc.OperatorBC(geomtime, lambda x, u, Bi_f: dde.grad.jacobian(u, x, i=0, j=1) + Bi_f * u, boundary_u)

### Initial condition: u(x,y,0) = 1
def initial_func(x):
    return np.ones((len(x), 1), dtype=np.float32)

### -------------------------------
data = dde.data.TimePDE(
    geomtime,
    pde=pde,
    ic_bcs=[bc_l, bc_r, bc_d, bc_u],
    num_domain=1000,
    num_boundary=500,
    num_test=500,
    train_distribution="pseudo",
)

### -------------------------------
degree = 3
func_space = dde.data.PowerSeries(N=degree + 1)
x_vals = np.linspace(0, 1, 60)[:, None]  # dummy BIF discretization
pde_op = dde.data.PDEOperator(
    data,
    func_space,
    x_vals,
    num_function=500,
    function_variables=[0],
)

### -------------------------------
neurons = 64
net = dde.nn.DeepONet(
    [60] + [neurons]*3,  # branch
    [3] + [neurons]*3,   # trunk
    "tanh",
    "Glorot normal"
)

### Output transform to enforce initial condition
net.apply_output_transform(lambda x, u: tf.ones_like(x[1][:,2:3]) + x[1][:,2:3]*u)

### -------------------------------
model = dde.Model(pde_op, net)
model.compile("adam", lr=1e-3, loss_weights=[1.0, 1.0, 1.0, 1.0, 10.0])  # Emphasize bc_up

### -------------------------------
losshistory, train_state = model.train(iterations=1000, batch_size=32)

### -------------------------------
model.save("pino_model", verbose=1)
dde.utils.external.plot_loss_history(losshistory, "pino_loss.png")

Despite following this approach, the **predictions are not accurate**, and the errors are high. I suspect there may be an **issue in my implementation or formulation**, particularly in the **operator representation or training setup**.

Could you please advise:

1. Whether this **implementation approach is correct** for physics-informed operator learning in DeepXDE
2. Any common pitfalls or best practices for handling **operator inputs like BIF**
3. Suggestions to improve training or reduce errors (e.g., function space, sampling, loss weighting)

Thank you very much for your time and guidance.

Best regards,
**Jariful Hassan**
Project Associate
Department of Mechanical Engineering
Indian Institute of Technology Palakkad

**Comment by echen5503** [comment]
Sorry for the late response. I am not an expert in DeepXDE, but looking at your code, here is what I think.

Q1. The specification seems correct, assuming you want: Entire heat field is 1 (heated) at the start, left and bottom are insulated, right side with constant convection strength, and the top has variable convection strength (what you're training your operator against).

Q2.
1. Because you use the laplacian (second derivative) in your PDE, don't use an activation that doesn't have a second derivative (like ReLU). In general, it is safe to use an activation that is infinitely differentiable (tanh).
2. Note that tensor flow does not support LFBGS (a second-derivative based optimizer meant to fine-tune model at the end). Try using pytorch if you plan to use LFBGS. (See https://github.com/lululxvi/deepxde/issues/2040)

Q2/Q3.
If you aren't getting good results with the code, it is probably attributable to some of these issues:
1. Iterations are not sufficient. Operator learning typically requires 10,000+ iterations. Fine-tuning with LFBGS can also help.
2. Watch the loss weightings for boundary conditions. If the top and right side have high biot numbers, then the gradient there will be steep, causing the optimizer to prioritize there. Considering that your loss weighting already weighs the top highly, this can cause the Neumann boundary conditions to be neglected. You could try learning rate annealing if you're not sure.
4. The top function space seems suspicious. In normal physical scenarios, one convection area is usually approximately uniform. Having 60 different points of different convection strength is very difficult to learn. If this is what you intend, increase the model size as the model is probably not complex enough to handle this.

**Comment by JarifulHassan** [comment]
Thank you for your kind and detailed response. I really appreciate your suggestions.
I have already addressed these points in my implementation. The problem required a large number of sampling points and functions in the function space because the top boundary has spatially varying convection coefficients. To handle this, I implemented the model using PyTorch with batching and increased the number of training samples and functions.
After these modifications, the model is now predicting the solution very well with good accuracy.
Thank you again for your helpful feedback.
Best regards,
Jariful Hassan

________________________________
